File: Depth-first_search(DFS)/c++/05_2048(Easy).cpp
// fast campus 강의
// https://www.acmicpc.net/problem/12100
// 2

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int N;

vector<vector<int>> rotate90(vector<vector<int>> b)
{
	vector<vector<int>> new_board(N, vector<int>(N, 0));
	for (int i = 0; i < N; i++)
		for (int j = 0; j < N; j++)
			new_board[j][N - i - 1] = b[i][j];
	return new_board;
}

vector<int> convert(vector<int> lst)
{
	vector<int> new_list1, new_list2;
	for (auto x : lst)
		if (x)
			new_list1.push_back(x);
	for (int i = 1; i < new_list1.size(); i++) {
		if (new_list1[i - 1] == new_list1[i]) {
			new_list1[i - 1] *= 2;
			new_list1[i] = 0;
		}
	}

	for (auto x : new_list1)
		if (x)
			new_list2.push_back(x);
	for (int i = 0; new_list2.size() != N; i++)
		new_list2.push_back(0);
	return new_list2;
}

int dfs(vector<vector<int>> b, int cnt)
{
	int ret = 0;
	for (int i = 0; i < N * N; i++)
		ret = max(ret, b[i / N][i % N]);
	if (cnt == 5)
		return ret;

	for (int i = 0; i < 4; i++) {
		vector<vector<int>> x(N);
		for (int j = 0; j < N; j++)
			x[j] = convert(b[j]);
		if (b != x)
			ret = max(ret, dfs(x, cnt + 1));
		b = rotate90(b);
	}
	return ret;
}

int main()
{
	cin >> N;
	vector<vector<int>> b(N, vector<int>(N, 0));
	for (int i = 0; i < N; i++)
		for (int j = 0; j < N; j++)
			cin >> b[i][j];

	cout << dfs(b, 0);
}

File: Depth-first_search(DFS)/python/05_2048(Easy).py
# fast campus 강의
# https://www.acmicpc.net/problem/12100
# 1

from copy import deepcopy

N = int(input())
board = [list(map(int, input().split())) for i in range(N)]

def rotate90(board, N):
    rotate_board = deepcopy(board)
    for i in range(N):
        for j in range(N):
            rotate_board[j][N - i - 1] = board[i][j]
    return rotate_board

def convert(lst, N):
    new_list = [i for i in lst if i != 0]
    for i in range(1, len(new_list)):
        if new_list[i - 1] == new_list[i]:
            new_list[i - 1] *= 2
            new_list[i] = 0
    new_list = [i for i in new_list if i != 0]
    return new_list + [0] * (N - len(new_list))

def dfs(N, board, count):
    ret = max([max(i) for i in board])
    if count == 0:
        return ret
    for _ in range(4):
        X = [convert(lst, N) for lst in board]
        if X != board:
            ret = max(ret, dfs(N, X, count - 1))
        board = rotate90(board, N)
    return ret

print(dfs(N, board, 5))